enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -1179,14 +1190,56 @@ func (dc *writeCloser) Write(p []byte) (int, error) {
 	return n, nil
 }
 
-// TODO(Maggie): Compress all data above a given size before writing it
+// zstdCompressor compresses bytes before writing them to the nested writer
+type zstdCompressor struct {
+	*ioutil.CustomCommitWriteCloser
+	compressBuf []byte
+	bufferPool  *bytebufferpool.Pool
+}
+
+func (z *zstdCompressor) Write(decompressedBytes []byte) (int, error) {
+	z.compressBuf = compression.CompressZstd(z.compressBuf, decompressedBytes)
+	compressedBytesWritten, err := z.CustomCommitWriteCloser.Write(z.compressBuf)
+	if err != nil {
+		return 0, err
+	}
+
+	metrics.CompressionRatio.With(prometheus.Labels{metrics.CompressionType: "zstd"}).Observe(float64(compressedBytesWritten) / float64(len(decompressedBytes)))
+	if compressedBytesWritten > len(decompressedBytes) {
+		metrics.BadCompressionBufferSize.With(prometheus.Labels{metrics.CompressionType: "zstd"}).Observe(float64(len(decompressedBytes)))
+	}
```

Comment:
does it make sense to move this to the Close() method so this only happens once at the end of the write?

Reply:
Changed. Now the metrics will be on the stream level instead of the chunk level, which I think will be more useful as a metric